core/src/main/java/org/apache/accumulo/core/client/summary/CountingSummarizer.java

```diff
@@ -0,0 +1,302 @@
+   *
+   * @return a function that used to copy the counter object. This function is only used when the collector has never seen the counter object before. In this
+   *         case the collector needs to possibly copy the counter object before using as map key. The default implementation is the
+   *         {@link UnaryOperator#identity()} function.
+   */
+  protected UnaryOperator<K> copier() {
+    return UnaryOperator.identity();
+  }
+
+  @Override
+  public Collector collector(SummarizerConfiguration sc) {
+    init(sc);
+    return new Collector() {
+
+      // Map used for computing summary incrementally uses ByteSequence for key which is more efficient than converting String for each Key. The
+      // conversion to String is deferred until the summary is requested.
+
+      private Map<K,MutableLong> counters = new HashMap<>();
+      private long tooMany = 0;
+      private long tooLong = 0;
+      private long seen = 0;
+      private long emitted = 0;
+      private long deleted = 0;
+      private Converter<K> converter = converter();
+      private Function<K,String> encoder = encoder();
+      private UnaryOperator<K> copier = copier();
+
+      private void incrementCounter(K counter) {
+        emitted++;
+
+        MutableLong ml = counters.get(counter);
+        if (ml == null) {
+          if (counters.size() >= maxCounters) {
+            // no need to store this counter in the map and get() it... just use instance variable
+            tooMany++;
+          } else {
+            // we have never seen this key before, check if its too long
+            if (encoder.apply(counter).length() >= maxCounterKeyLen) {
+              tooLong++;
+            } else {
+              counters.put(copier.apply(counter), new MutableLong(1));
+            }
+          }
+        } else {
+          // using mutable long allows calling put() to be avoided
+          ml.increment();
+        }
+      }
+
+      @Override
+      public void accept(Key k, Value v) {
+        seen++;
+        if (ignoreDeletes && k.isDeleted()) {
+          deleted++;
+        } else {
+          converter.convert(k, v, this::incrementCounter);
+        }
+      }
+
+      @Override
+      public void summarize(StatisticConsumer sc) {
+        StringBuilder sb = new StringBuilder(COUNTER_STAT_PREFIX);
+
+        for (Entry<K,MutableLong> entry : counters.entrySet()) {
+          sb.setLength(COUNTER_STAT_PREFIX.length());
+          sb.append(encoder.apply(entry.getKey()));
+          sc.accept(sb.toString(), entry.getValue().longValue());
+        }
+
+        sc.accept(TOO_MANY_STAT, tooMany);
+        sc.accept(TOO_LONG_STAT, tooLong);
+        sc.accept(EMITTED_STAT, emitted);
+        sc.accept(SEEN_STAT, seen);
+        sc.accept(DELETES_IGNORED_STAT, deleted);
+      }
+    };
+  }
+
+  @Override
+  public Combiner combiner(SummarizerConfiguration sc) {
+    init(sc);
+    return new Combiner() {
```

Comment:
Make this a private static class instead of inline to make testing `merge` easier